compiler/core/lam_util.ml

```diff
@@ -38,86 +38,60 @@ let add_required_modules ( x : Ident.t list) (meta : Lam_stats.t) =
-  | _, _, Lprim {primitive ; args =  [Lvar w]; loc ; _} when Ident.same w param 
-                                                          &&  (function | Lam_primitive.Pmakeblock _ -> false | _ ->  true) primitive
-    (* don't inline inside a block *)
-    ->  Lam.prim ~primitive ~args:[arg]  loc 
-  (* we can not do this substitution when capttured *)
-  (* | _, Lvar _, _ -> (\** let u = h in xxx*\) *)
-  (*     (\* assert false *\) *)
-  (*     Ext_log.err "@[substitution >> @]@."; *)
-  (*     let v= subst_lambda (Map_ident.singleton param arg ) l in *)
-  (*     Ext_log.err "@[substitution << @]@."; *)
-  (* v *)
-  | _, _, Lapply {ap_func=fn; ap_args = [Lvar w]; ap_info; ap_transformed_jsx} when
-      Ident.same w param &&
-      (not (Lam_hit.hit_variable param fn ))
-    -> 
-    (* does not work for multiple args since 
-        evaluation order unspecified, does not apply 
-        for [js] in general, since the scope of js ir is loosen
-
-        here we remove the definition of [param]
-       {[ let k = v in (body) k 
-       ]}
-        #1667 make sure body does not hit k 
-    *)
-    Lam.apply fn [arg] ap_info ~ap_transformed_jsx
-  | (Strict | StrictOpt ),
-    ( Lvar _    | Lconst  _ | 
-      Lprim {primitive = Pfield (_ , Fld_module _) ;  
-             args = [ Lglobal_module _ | Lvar _ ]; _}) , _ ->
-    (* (match arg with  *)
-    (* | Lconst _ ->  *)
-    (*     Ext_log.err "@[%a %s@]@."  *)
-    (*       Ident.print param (string_of_lambda arg) *)
-    (* | _ -> ()); *)
-    (* No side effect and does not depend on store,
-        since function evaluation is always delayed
-    *)
-    Lam.let_ Alias param arg l
-  | ( (Strict | StrictOpt ) ), (Lfunction _ ), _ ->
-    (*It can be promoted to [Alias], however, 
-        we don't want to do this, since we don't want the 
-        function to be inlined to a block, for example
-      {[
-        let f = fun _ -> 1 in
-        [0, f]
-      ]}
-        TODO: punish inliner to inline functions 
-        into a block 
-    *)
-    Lam.let_ StrictOpt  param arg l
-  (* Not the case, the block itself can have side effects 
-      we can apply [no_side_effects] pass 
-      | Some Strict, Lprim(Pmakeblock (_,_,Immutable),_) ->  
-        Llet(StrictOpt, param, arg, l) 
-  *)      
-  | Strict, _ ,_  when Lam_analysis.no_side_effects arg ->
-    Lam.let_ StrictOpt param arg l
-  | Variable, _, _ -> 
-    Lam.let_ Variable  param arg l
-  | kind, _, _ -> 
-    Lam.let_ kind  param arg l
-(* | None , _, _ -> 
-   Lam.let_ Strict param arg  l *)
+(* refine_let normalises let-bindings so we avoid redundant locals while
+   preserving evaluation semantics. *)
+let refine_let ~kind param (arg : Lam.t) (l : Lam.t) : Lam.t =
+  let is_block_constructor = function
+    | Lam_primitive.Pmakeblock _ -> true
+    | _ -> false
+  in
+  let is_alias_candidate (lam : Lam.t) =
+    match lam with
+    | Lvar _ | Lconst _ -> true
+    | Lprim { primitive = Psome_not_nest; _ } -> true
```

Comment:
Oh here it is. The change.
Please add comments to explain this function.

Reply:
I suspect there's a semantic property that all these cases have in common, and such property justifies them being an alias candidate (btw the name "alias candidate" might or might be too vague of a name, not sure).